central/graphql/resolvers/image_components.go

```diff
@@ -0,0 +1,115 @@
+package resolvers
+
+import (
+	"context"
+	"time"
+
+	"github.com/graph-gophers/graphql-go"
+	"github.com/pkg/errors"
+	"github.com/stackrox/rox/central/metrics"
+	"github.com/stackrox/rox/pkg/features"
+	pkgMetrics "github.com/stackrox/rox/pkg/metrics"
+	"github.com/stackrox/rox/pkg/search"
+	"github.com/stackrox/rox/pkg/utils"
+)
+
+func init() {
+	schema := getBuilder()
+	utils.Must(
+		// NOTE: This list is and should remain alphabetically ordered
+		schema.AddExtraResolvers("ImageComponent", []string{
+			"activeState(query: String): ActiveState",
+			"deploymentCount(query: String, scopeQuery: String): Int!",
+			"deployments(query: String, scopeQuery: String, pagination: Pagination): [Deployment!]!",
+			"fixedIn: String! @deprecated(reason: \"use 'fixedBy'\")",
+			"imageCount(query: String, scopeQuery: String): Int!",
+			"images(query: String, scopeQuery: String, pagination: Pagination): [Image!]!",
+			"imageVulnerabilityCount(query: String, scopeQuery: String): Int!",
+			"imageVulnerabilityCounter(query: String): VulnerabilityCounter!",
+			"imageVulnerabilities(query: String, scopeQuery: String, pagination: Pagination): [ImageVulnerability]!",
+			"lastScanned: Time",
+			"location(query: String): String!",
+			"topImageVulnerability: ImageVulnerability",
+			"unusedVarSink(query: String): Int",
+
+			"plottedVulns(query: String): PlottedVulnerabilities!", // TODO
+		}),
+		schema.AddQuery("imageComponent(id: ID): ImageComponent"),
+		schema.AddQuery("imageComponents(query: String, scopeQuery: String, pagination: Pagination): [ImageComponent!]!"),
+		schema.AddQuery("imageComponentCount(query: String): Int!"),
+
+		// TODO
+		schema.AddExtraResolver("ImageScan", `components(query: String, pagination: Pagination): [EmbeddedImageScanComponent!]!`),
+		schema.AddExtraResolver("ImageScan", `componentCount(query: String): Int!`),
+	)
+}
+
+// ImageComponentResolver represents a generic resolver of image component fields.
+// Values may come from either an embedded component context, or a top level component context.
+// NOTE: This list is and should remain alphabetically ordered
+type ImageComponentResolver interface {
+	ActiveState(ctx context.Context, args RawQuery) (*activeStateResolver, error)
+	DeploymentCount(ctx context.Context, args RawQuery) (int32, error)
+	Deployments(ctx context.Context, args PaginatedQuery) ([]*deploymentResolver, error)
+	FixedIn(ctx context.Context) string
+	FixedBy(ctx context.Context) string
+	ID(ctx context.Context) graphql.ID
+	ImageCount(ctx context.Context, args RawQuery) (int32, error)
+	Images(ctx context.Context, args PaginatedQuery) ([]*imageResolver, error)
+	ImageVulnerabilityCount(ctx context.Context, args RawQuery) (int32, error)
+	ImageVulnerabilityCounter(ctx context.Context, args RawQuery) (*VulnerabilityCounterResolver, error)
+	ImageVulnerabilities(ctx context.Context, args PaginatedQuery) ([]ImageVulnerabilityResolver, error)
+	LastScanned(ctx context.Context) (*graphql.Time, error)
```

Comment:
LastScanned() is named NodeComponentLastScanned() for node components to separate out the implementations. Should this be renamed to ImageComponentLastScanned() as well?

Reply:
`LastScanned` is a field on components. We should not rename it. It is always wrt to **one component**.